File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m,n=len(grid),len(grid[0])

        one=[]
        visited=[[-1 for _ in range(n)]for _ in range(m)]
        for i in range(m):
            for j in range(n):
                if grid[i][j]=="1":
                    one.append((i,j))
        

        def dfs(r,c):
            visited[r][c]=1
            dir=[(-1,0),(1,0),(0,1),(0,-1)]

            for dr,dc in dir:
                xr=r+dr
                xc=c+dc

                if xr>=0 and xr<m and xc>=0 and xc<n and visited[xr][xc]==-1 and grid[xr][xc]=="1":
                    dfs(xr,xc)
        islands=0

        for x,y in one:
            if visited[x][y]==-1:
                islands+=1
                dfs(x,y)

        return islands
        

File: 200-number-of-islands/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int m=grid.length;
        int n=grid[0].length;
        int[][] visited=new int[m][n];

        for(int i=0;i<m;i++){
            Arrays.fill(visited[i],-1);
        }
        int count=0;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(visited[i][j]==-1 && grid[i][j]=='1'){
                    count++;
                    dfs(grid,visited,i,j,m,n);
                }
            }
        }
        return count;
    }
    public void dfs(char[][] grid,int[][] visited,int x,int y,int m,int n){
        visited[x][y]=1;
        int[] xdir={-1,1,0,0};
        int[] ydir={0,0,-1,1};

        for(int i=0;i<4;i++){
            int r=x+xdir[i];
            int c=y+ydir[i];

            if(r>=0 && r<m && c>=0 && c<n && visited[r][c]==-1 && grid[r][c]=='1'){
                dfs(grid,visited,r,c,m,n);
            }
        }
    }
}